usr/share/rear/build/default/500_patch_sshd_config.sh

```diff
@@ -1,33 +1,102 @@
+# which has to be enabled in the original system to work in the recovery system.
+# The funny [] around a letter makes 'shopt -s nullglob' remove this file from the list if it does not exist.
+# Files without a [] are mandatory.
+local sshd_config_files=( $ROOTFS_DIR/etc/ssh/sshd_co[n]fig $ROOTFS_DIR/etc/sshd_co[n]fig $ROOTFS_DIR/etc/openssh/sshd_co[n]fig )
+if test "${sshd_config_files[*]}" ; then
+    sed -i -e 's/ChallengeResponseAuthentication.*/ChallengeResponseAuthentication no/ig' \
+           -e 's/UsePAM.*/UsePam no/ig' \
+           -e 's/ListenAddress.*/ListenAddress 0.0.0.0/ig' \
+           -e '1i\PrintMotd no' \
+        ${sshd_config_files[@]}
+    # Allow password authentication in the recovery system only if SSH_ROOT_PASSWORD is specified:
+    if test "$SSH_ROOT_PASSWORD" ; then
+        sed -i -e 's/PasswordAuthentication.*/PasswordAuthentication yes/ig' ${sshd_config_files[@]}
+        sed -i -e 's/PermitRootLogin.*/PermitRootLogin yes/ig' ${sshd_config_files[@]}
     else
-        sed -i  -e 's/PasswordAuthentication.*/PasswordAuthentication no/ig' ${SSH_CONFIG_FILES[@]}
+        sed -i  -e 's/PasswordAuthentication.*/PasswordAuthentication no/ig' ${sshd_config_files[@]}
     fi
 else
-    Log "WARNING: sshd configuration file missing"
+    LogPrintError "No sshd configuration files"
+fi
+
+# Generate new SSH host key in the recovery system when no SSH host key file
+# had been copied into the the recovery system in rescue/default/500_ssh.sh
+# cf. https://github.com/rear/rear/issues/1512#issuecomment-331638066
+# In SLES12 "man ssh-keygen" reads:
+#   -t dsa | ecdsa | ed25519 | rsa | rsa1
+#      Specifies the type of key to create.
+#      The possible values are "rsa1" for protocol version 1
+#      and "dsa", "ecdsa", "ed25519", or "rsa" for protocol version 2.
+# The above GitHub issue comment proposes a static
+#   ssh-keygen -t ed25519 -N '' -f "..."
+# but the key type ed25519 is not supported in older systems like SLES11.
+# On SLES10 "man ssh-keygen" reads:
+#   -t type
+#      Specifies the type of key to create.
+#      The possible values are rsa1 for protocol version 1
+#      and rsa or dsa for protocol version 2.
+# Currently (October 2017) ReaR is kept working on older systems
+# like SLES10 cf. https://github.com/rear/rear/issues/1522
+# and currently this backward compatibility should not be broken
+# (for the future see https://github.com/rear/rear/issues/1390)
+# so that we try to generate all possible types of keys.
+# This is in compliance what there is on a default SLES system, e.g.
+# on a default SLES10 there is
+#   /etc/ssh/ssh_host_rsa_key
+#   /etc/ssh/ssh_host_dsa_key
+# on a default SLES11 there is additionally
+#   /etc/ssh/ssh_host_ecdsa_key
+# on a default SLES12 there is additionally
+#   /etc/ssh/ssh_host_ed25519_key
+# only a rsa1 type key does not exists so that rsa1 is also not generated here:
+local ssh_host_key_types="rsa dsa ecdsa ed25519"
```

Comment:
sshd works with just one key, and for login purposes, one key is sufficient.

To preserve ReaR backward compatibility starting with SSH protocol version 2 (the default since 2001), it should be sufficient to just generate an "rsa" key. "ed25519" would provide better long-term security, but this is probably not necessary on a rescue system, which is used at most once.

So the whole code could be simplified by checking if _any_ key is present, and, if not, just generate an "rsa" key.

Reply:
Guess what I had implemented initially ;-)
but then I found out that on a default SLE system
I have all those key types (as listed in my comment
in the code).

Because I am not a SSH expert I can only trust
our SUSE SSH maintainers that they "do it right"
(for whatever reasons that I cannot know) so that
I simply did the same to get the recovery system
compatible with an original SLES system to avoid
unexpected failures during recovery in whatever special
cases that I cannot foresee.

Reply:
Two points:
1. It is no longer recommended to use host keys other than RSA and ED25519. See section 2.1 in  [BetterCrypto.org Applied Crypto Hardening](https://bettercrypto.org/static/applied-crypto-hardening.pdf).
2. Generating old-style host keys will only help with ssh clients which are more than 16 years old (OpenSSH versions < 2.9, see [OpenSSH: Release Notes](https://www.openssh.com/releasenotes.html)).

Reply:
I will only generate the particular type of key if
/etc/ssh/ssh_host_TYPE_key
exists in the original system.
This keeps compatibility with what there is in the original system
and should fulfill what is requested in
https://github.com/rear/rear/pull/1530#discussion_r143948453

Reply:
FYI regarding very old ssh clients:
On my SLES10 SP4 system I have
<pre>
# rpm -qa | grep -i ssh
openssh-5.1p1-41.8.20
openssh-askpass-5.1p1-41.8.20

# ssh -V
OpenSSH_5.1p1, OpenSSL 0.9.8a 11 Oct 2005
</pre>

Reply:
So your OpenSSH is a child of 2008! Not too bad. ;-)

Reply:
For our "latest greatest" see
https://build.opensuse.org/package/show/network/openssh

Reply:
Now I generate a single rsa key as fallback in the recovery system
if no (other) key is there (e.g. if there is no rsa key on the original
system and generating all other key types had failed).
This makes ssh access to the recovery system work for me
even when I removed all /etc/ssh/*key* files on the original system.
Because this could make the recovery system "more accessible"
from remote compared to the original system I report when
the fallback rsa key is generated via LogPrint (i.e. only in
verbose mode that information is shown on the users terminal).